Limited Wasp version to ^16.0.0 && <=16.15.0

Martinsos opened the pull request:
I basically added condition <=16.15.0 since 16.15.1 is causing us problems with Prisma (https://github.com/wasp-lang/wasp/issues/631).

I also removed .nvmrc files from the generated project because I concluded they are not useful.
I also refactored SemanticVersion to follow npm specification that Filip shared with me.

Once we merge this, we can release the new version of Wasp that should then warn users if they try to use node 16.15.1 or a later version.

Martinsos commented:
TODO: I should also update the docs, but I am not sure how to best update them hm.
I am talking about the getting-started.md in our web/docs. It mentions 16 on many places. Should I just add one special note that says 16.15.1 does not work? Or should I change it completely to use a specific version of node, like 16.15.0? Not sure.

Martinsos commented on waspc/test/SemanticVersionTest.hs:
```diff
@@ -1,39 +1,91 @@
+    it "Recognizes only version v to be in range '=v'" $
+      testRange
+        (rangeFromVersion (Equal, Version 1 2 3))
+        [ ((0, 5, 5), False),
```
Thank you! I find it fun in Haskell to rewrite the tests with these helpers, somehow it is really convenient to do that, since everything is an expression.

sodic replied:
I agree, this looks really nice!

Martinsos commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -18,24 +27,51 @@ data Version = Version
+instance Show Comparator where
+  show (Comparator op v) = show op ++ show v
+
+data ComparatorSet = ComparatorSet (NonEmpty Comparator)
```
Ha thanks :D!

Martinsos commented:
> Great job @Martinsos! I don't have anything to add here really, it all looks good and makes sense. I tested it out locally and it looks good as well. @sodic may be interested in giving it a look since he added SV, but this is good to go in my book! rocket

Thank you!!

I worried a bit about making SemanticVersion more complicated than before, but on the other hand now it follows node semver spec, so at least we are not making up the concepts / names any more. Even if the ones they provide are a bit clunky regarding naming (comparator, really).

I also worried for a moment when I realized this is not a general SemVer spec anymore, instead it is one tied to node / npm, but on the other hand pure SemVer is not so useful and in practice node / npm is very used, so I believe this is ok.

Oh btw I used a Semigroup, I think this is first time I implemented Semigroup instance :D! Very simple but fun :D.

One unusual thing I also did are those helper functions at the bottom of the SemanticVersion file. I did those because I imagined that in most cases, you don't want to construct the Range with all the steps it needs -> ComparatorSet, Comparator. So I created these to "cheat" a bit and allow you to deal with constructing Ranges without knowing about all that intermittent structure. But you can still use all the structure if you need it. It feels a tiny bit hacky due to naming -> what does it mean to construct range from version -> but in practice it felt ok. Hm.

Also, I removed Prisma version Range at the end because I realized that at some places we use it with assumption that is an exact version and not range, which sounded dangerous. So I concluded, if we are assuming it is an exact version anyway, let's use direct version directly.

shayneczyzewski commented:
> > Great job @Martinsos! I don't have anything to add here really, it all looks good and makes sense. I tested it out locally and it looks good as well. @sodic may be interested in giving it a look since he added SV, but this is good to go in my book! rocket
> 
> Thank you!!
> 
> I worried a bit about making SemanticVersion more complicated than before, but on the other hand now it follows node semver spec, so at least we are not making up the concepts / names any more. Even if the ones they provide are a bit clunky regarding naming (comparator, really).
> 
> I also worried for a moment when I realized this is not a general SemVer spec anymore, instead it is one tied to node / npm, but on the other hand pure SemVer is not so useful and in practice node / npm is very used, so I believe this is ok.
> 
> Oh btw I used a Semigroup, I think this is first time I implemented Semigroup instance :D! Very simple but fun :D.
> 
> One unusual thing I also did are those helper functions at the bottom of the SemanticVersion file. I did those because I imagined that in most cases, you don't want to construct the Range with all the steps it needs -> ComparatorSet, Comparator. So I created these to "cheat" a bit and allow you to deal with constructing Ranges without knowing about all that intermittent structure. But you can still use all the structure if you need it. It feels a tiny bit hacky due to naming -> what does it mean to construct range from version -> but in practice it felt ok. Hm.
> 
> Also, I removed Prisma version Range at the end because I realized that at some places we use it with assumption that is an exact version and not range, which sounded dangerous. So I concluded, if we are assuming it is an exact version anyway, let's use direct version directly.
> 
> Also, one thing I did that is a bit of anti-pattern is rely on the implementation of `show` instance. Usually, the recommendation is to not do that, and to use `show` as debugging mechanism, and if you need to transform smth into String representation that follows a specific format (like we do here, we rely on `show` producing SemVer-recognized strings), that you should use specialized functions where that intention is clear and documented, either in a name or in a comment. That way you avoid somebody modifying `show` without knowing what the rest of the code expects from it and possibly breaking those expectations. But I thought, there isn't much other ways to `show` this piece of data, so maybe I can get by with it. Hm. Maybe I should change this though.

Yeah, that makes sense. I think even trending slightly from vanilla SV here is good since it suits our purposes well.

I noticed those there too... I spent like 10 min trying to golf `Range . pure . ComparatorSet . pure . uncurry Comparator` to see if I could somehow operate on `NonEmpty` a bit easier, but I could not lol. The actual usage of those made sense to me at least.

Ah, good point on `Show`. That best practice advice does make sense. I guess as long as it is backed up by a test it may be ok for now, but yeah, I can see that any "serialization" process may benefit from its own distinct function that is less likely to have accidental changes when someone changes that show behavior thinking it is no biggie. That may be one thing to consider.

sodic commented on waspc/test/SemanticVersionTest.hs:
```diff
@@ -1,39 +1,91 @@
+                    ]
+                )
+            )
+              <> rangeFromVersion (Equal, Version 0 5 6)
```
Haha, nice. Definitely justifies including a Semigroup :smile:

Martinsos replied:
hah yeah right :D?

sodic reviewed (commented):
Great job with extending `SemanticVersion`. This module truly lends itself to showcasing different features of Haskell, perhaps we can even task the students with something similar during the FER course next semester?

I agree with your comment on how it might have become a bit too complicated. I suggested a couple of different approaches in one of the comments.

One additional note: we'll most likely never reach the full capabilities of Node Semver unless we significantly change the implementation in favor of something "dirtier." The main reasons for this are those specifiers that could not work in terms of our type `Version` (e.g., `*`, `1.x`, `^1.x.x`) and depend on how the programmer specified the version string. Supporting these guys in a clean way would probably mean removing `Version` altogether. I talk more about this in the other comment.

With all that said, I'd definitely go ahead with merging this. I didn't find any errors in the PR, only potential design tweaks. Better to get it out and fix the problem for our users. We can deal with refactoring later if necessary, it's all internal code anyway.

sodic commented:
> Also, one thing I did that is a bit of anti-pattern is rely on the implementation of show instance.

I think this one is on me :upside_down_face: 

> Also, I removed Prisma version Range at the end because I realized that at some places we use it with assumption that is an exact version and not range, which sounded dangerous. So I concluded, if we are assuming it is an exact version anyway, let's use direct version directly.

I remember having a few question marks about the Prisma version when I was looking at this last time. Probably best to leave it at an exact version for now and minimize the chances of it causing problems for our users.

Martinsos commented:
> > Also, I removed Prisma version Range at the end because I realized that at some places we use it with assumption that is an exact version and not range, which sounded dangerous. So I concluded, if we are assuming it is an exact version anyway, let's use direct version directly.
> 
> Where did you see this? I'm guessing you're referring to [this line](https://github.com/wasp-lang/wasp/pull/634/files#diff-b24716d343af17daf33f7e342c7d16fc269f712a03efff342a583f52fd3b3ddfR27) and the comment above it. If that's the case, leaving the bounds should still be OK. The comment is a bit misleading, it should say "We want to pin down a _version range_." Unless `npm install prisma@1.2.3` will also happily install `1.2.4` and `1.2.5` and something else? According to the [API reference](https://github.com/wasp-lang/wasp/pull/634/files#diff-b24716d343af17daf33f7e342c7d16fc269f712a03efff342a583f52fd3b3ddfR27), that shouldn't be the case

Yup, that line! It didn't make sense to me that we say npx to install prisma version bounds - you have to give it specific version, right? So it made no sense to me to provide bounds there. But if that is so, and we know we are using specific prisma version (and we are!), what is the purpose of bounds at all? So I removed them.

sodic commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -24,31 +25,41 @@ data Version = Version
+-- | We rely on `show` here to produce valid semver representation of comparator set.
 instance Show ComparatorSet where
-  show (ComparatorSet comps) = unwords $ show <$> toList comps
+  show (ComparatorSet comps) = unwords $ show <$> NE.toList comps
```
And I was just about to applaud the lack of acronyms in this PR! :smile:

sodic commented:
> > > Also, I removed Prisma version Range at the end because I realized that at some places we use it with assumption that is an exact version and not range, which sounded dangerous. So I concluded, if we are assuming it is an exact version anyway, let's use direct version directly.
> > 
> > 
> > Where did you see this? I'm guessing you're referring to [this line](https://github.com/wasp-lang/wasp/pull/634/files#diff-b24716d343af17daf33f7e342c7d16fc269f712a03efff342a583f52fd3b3ddfR27) and the comment above it. If that's the case, leaving the bounds should still be OK. The comment is a bit misleading, it should say "We want to pin down a _version range_." Unless `npm install prisma@1.2.3` will also happily install `1.2.4` and `1.2.5` and something else? According to the [API reference](https://github.com/wasp-lang/wasp/pull/634/files#diff-b24716d343af17daf33f7e342c7d16fc269f712a03efff342a583f52fd3b3ddfR27), that shouldn't be the case
> 
> Yup, that line! It didn't make sense to me that we say npx to install prisma version bounds - you have to give it specific version, right? So it made no sense to me to provide bounds there. But if that is so, and we know we are using specific prisma version (and we are!), what is the purpose of bounds at all? So I removed them.

I somehow linked the wrong thing. This is what "API reference" was supposed to link to: https://docs.npmjs.com/cli/v8/commands/npm-install. I've edited both previous comments to fix it.

Martinsos commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -18,24 +27,51 @@ data Version = Version
+data Comparator = Comparator Operator Version
+
+instance Show Comparator where
+  show (Comparator op v) = show op ++ show v
```
Awesome analysis @sodic, thanks for this!

From their spec, I got the idea that they use Version - Operator - Comparator - ComparatorSet - Range as a basic structure, and then the rest is sugar on top of it (they do call it "sugar" actually).
So I followed the same approach, and decided this will be the basis, and we will add other special stuff, like special ranges, comparators and operators on top of that, somehow.
I did the exception of adding `^` directly as the primitive operator though, because I saw no reason why that wouldn't work, and I didn't want to further complicate things for now. But I agree, we should take that one out, since it is a special comparator set that desugarizes into the primitive comparators.

I think you are voting for (1) :D, but I would suggest we go with (3) + a TODO for `^`, since that will allow us to merge it right now + I think it is a good foundation for further expanding this code as needed. While I agree it might be a bit overly complex, a big benefit is that we can just copy what they did without making up our stuff. We could reduce the complexity, but then when we decide to add something we might uncover that we need more structure and will likely again end up with what they did, or with something else that has different names and maybe structure but is similarly complex, assuming their solution is relatively good / not far away from minimal required complexity.

sodic replied:
Makes sense, we merged it and agreed on future steps (written in todos).

Martinsos commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -18,24 +27,51 @@ data Version = Version
-instance Show VersionBounds where
-  show (BackwardsCompatibleWith version) = "^" ++ show version
-  show (Exact version) = show version
+data Range = Range (NonEmpty ComparatorSet)
```
Awesome, did it :D!

Martinsos commented:
Ok so what I did:
- left SemanticVersion how it is but added TODO for ^ and added Monoid for Range.
- left only PrismaVersion, no PrismaVersionRange, due to concluding that range wouldn't work the way it was currently provided to shell command due to missing apostrophes and I didn't want to spend time now figuring that out and testing -> once somebody will need prisma version to be range, they can do it.
- added docs
- added tests

This should be enough to ship I think!

matijaSos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -12,21 +11,30 @@ import qualified Wasp.SemanticVersion as SV
+      (SV.LessThanOrEqual, latestLTSExactVersionThatWeKnowWorks)
+    ]
+  where
+    latestLTSVersion = SV.Version 16 0 0
```
maybe rename to `latestLTSNodeVersion`? technically not needed since it is localized, it's just that right below we have the same name but meant for npm, so it might be a bit confusing, this way we make sure the reader knows what it is about

matijaSos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -12,21 +11,30 @@ import qualified Wasp.SemanticVersion as SV
+    ]
+  where
+    latestLTSVersion = SV.Version 16 0 0
+    -- There is a bug in node 16.15.1 (more correctly, in npm 8.11.0 that comes with it)
```
Should this be a NOTE?

matijaSos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -12,21 +11,30 @@ import qualified Wasp.SemanticVersion as SV
+      (SV.LessThanOrEqual, latestLTSExactVersionThatWeKnowWorks)
+    ]
+  where
+    latestLTSVersion = SV.Version 8 0 0 -- Goes with node 16
```
what does it mean to go with? No other combo is ok? do we check that the right combo is used?

matijaSos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -12,21 +11,30 @@ import qualified Wasp.SemanticVersion as SV
+    latestLTSVersion = SV.Version 8 0 0 -- Goes with node 16
+    latestLTSExactVersionThatWeKnowWorks = SV.Version 8 5 5 -- Goes with node 16.15.0
 
 prismaVersion :: SV.Version
```
then this should be `supportedPrismaVersion` hmm. Maybe `Common.hs` is not the best (for sure it isn't, it shouldn't exist according to uncle bob :D) place for this, we could extract version stuff under e.g. `SupportedVersions.hs`?

matijaSos commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -1,41 +1,104 @@
   }
   deriving (Eq, Ord)
 
+-- | We rely on `show` here to produce valid semver representation of version.
```
not really here, but in all the other places but here, right? Maybe add a test (if we don't have it yet) to ensure that is the fact?

matijaSos commented on waspc/src/Wasp/SemanticVersion.hs:
```diff
@@ -1,41 +1,104 @@
+data ComparatorSet = ComparatorSet (NonEmpty Comparator)
+  deriving (Eq)
+
+-- | We rely on `show` here to produce valid semver representation of comparator set.
```
also not here

matijaSos commented on waspc/test/SemanticVersionTest.hs:
```diff
@@ -1,39 +1,133 @@
+            ((2, 0, 0), False)
+          ]
+  where
+    testRange :: Range -> [((Natural, Natural, Natural), Bool)] -> Expectation
```
maybe something bit more verbose, e.g. `testIfInRange`, or `testIfVersionsInRange`

matijaSos commented on web/docs/getting-started.md:
```diff
@@ -53,6 +53,17 @@ We recommend using [nvm](https://github.com/nvm-sh/nvm) for managing your Node.j
+
+Why does Wasp require this specific `node` range and doesn't support a newer version x.y.z?
+
+At Wasp, we focus on supporting the latest LTS ("long-term-support") Node.js version, since it guarantees stability and active maintainance, which is why official Node.js team recommends it for usage in production.
```
"for usage in production" -> "using it in production" - maybe that is nicer/simpler?

matijaSos commented on web/docs/getting-started.md:
```diff
@@ -53,6 +53,17 @@ We recommend using [nvm](https://github.com/nvm-sh/nvm) for managing your Node.j
+Why does Wasp require this specific `node` range and doesn't support a newer version x.y.z?
+
+At Wasp, we focus on supporting the latest LTS ("long-term-support") Node.js version, since it guarantees stability and active maintainance, which is why official Node.js team recommends it for usage in production.
+Therefore, a specific Wasp release will usually require the version of Node.js that was LTS at that point of time.
```
should be "point in time" I think - but you can just say "at that time"

matijaSos commented on web/docs/getting-started.md:
```diff
@@ -53,6 +53,17 @@ We recommend using [nvm](https://github.com/nvm-sh/nvm) for managing your Node.j
+Therefore, a specific Wasp release will usually require the version of Node.js that was LTS at that point of time.
+Check out https://nodejs.org/en/about/releases/ for more details about Node.js releases.
+
+Sometimes we will make an exception to that and additionaly limit the Node.js version or postpone switching to the latest LTS if there are certain issues with new Node.js version, in which case we will catch up once those are resolved on Node.js side or we find a workaround on Wasp side.
```
I think you can remove "to that"

matijaSos reviewed (commented):
very cool!
